integrations/tensorflow/bindings/python/pyiree/tf/support/tf_utils.py

```diff
@@ -60,27 +60,42 @@ def to_mlir_type(dtype: np.dtype) -> str:
     raise TypeError(f"Expected integer or floating type, but got {dtype}")
 
 
-def get_shape_and_dtype(array: np.ndarray,
+def get_shape_and_dtype(value: Union[int, float, np.ndarray],
                         allow_non_mlir_dtype: bool = False) -> str:
-  shape_dtype = [str(dim) for dim in list(array.shape)]
-  if np.issubdtype(array.dtype, np.number):
-    shape_dtype.append(to_mlir_type(array.dtype))
-  elif allow_non_mlir_dtype:
-    shape_dtype.append(f"<dtype '{array.dtype}'>")
+  """Returns a string that denotes the shape and type of value."""
+  if not isinstance(value, (int, float, np.ndarray)):
+    raise TypeError("Expected 'value' to be one of 'int', 'float' or "
+                    f"'np.ndarray' but got {type(value)}.")
+
+  if isinstance(value, np.ndarray):
+    shape_dtype = [str(dim) for dim in list(value.shape)]
+    if np.issubdtype(value.dtype, np.number):
+      shape_dtype.append(numpy_to_mlir_type(value.dtype))
+    elif allow_non_mlir_dtype:
+      shape_dtype.append(f"<dtype '{value.dtype}'>")
+    else:
+      raise TypeError(
+          f"Expected integer or floating type, but got {value.dtype}")
+    return "x".join(shape_dtype)
   else:
-    raise TypeError(f"Expected integer or floating type, but got {array.dtype}")
-  return "x".join(shape_dtype)
+    return "f32" if isinstance(value, float) else "i32"
+
+
+def to_iree_data_string(value: Union[int, float, np.ndarray]):
+  if not isinstance(value, (int, float, np.ndarray)):
+    raise TypeError("Expected 'value' to be one of 'int', 'float' or "
+                    f"'np.ndarray' but got {type(value)}.")
+
+  shape_dtype = get_shape_and_dtype(value, allow_non_mlir_dtype=True)
+  if isinstance(value, np.ndarray):
+    value = " ".join([str(x) for x in value.flatten()])
+  return f"{shape_dtype}={value}"
```

Comment:
The original numpy arguments and datatype are not necessarily going to be the same as those passed to the underlying IREE function (e.g. an f16 in numpy will not be passed to IREE as an f16). I recommend that instead of implementing this in python, you do this through the C++ bindings. Get the buffers from the packed inputs and call iree_hal_buffer_view_format 

https://github.com/google/iree/blob/39479aa59b7d70cb8fc7abb9ec9338700366ec95/iree/hal/api.h#L892

on it. This means we only have to solve any data type conversion logic in one place and avoid having two parallel implementations of this printing/parsing logic (which was one of the things Ben was advising we avoid).

Reply:
@benvanik to confirm I'm not quoting him wrong :-P

Reply:
This is mostly just refactoring what Hanhan wrote to separate out the saving logic. The only functional change is that `int` and `float` types get parsed to `i32` and `f32` respectively. Parsing is not intended to be supported.

We could still go through C++ for the flagfile inputs if we need to do something special to the types, but I thought the `vmfb` itself was what determined how the types should be read, and what we print here is only used for a coherence check? And I imagine we shouldn't have to worry about weird float formats if we're printing them.

Reply:
I'm fine with this for now, but I do think we should consider doing this via the C++ instead in a follow up

Reply:
Capturing this at the C++ level would be great, and I would consider it necessary unless we had CI that unit tested this code against whatever is going to be reading the values on the other side.

Reply:
I added a function for this in the C++ bindings and routed it up to the `CompiledModule` level through `BoundFunction`, which allows us to also serialize the outputs. (The C++ function can't be generally used to convert numpy arrays as a `abi` instance is required).